devnotes/isteps/14_DRAM_init/1_mss_memdiag.md

```diff
@@ -0,0 +1,633 @@
+                fapi2::buffer<uint64_t> l_value_second = i_invert ? ~l_cache_line.second : l_cache_line.second;
+                fapi2::putScom(p, PT::RMW_WRT_BUF_DATA_REG, l_value_first);
+                // In order for the data to actually be written into the RMW buffer, we must issue a putscom to the MCA_AAER register
+                // This register is used for the ECC, we will just write all zero to this register.  The ECC will be auto generated
+                // when the aacr MCA_WREITE_AACR_ECCGEN bit is set
+                fapi2::putScom(p, PT::RMW_WRT_BUF_ECC_REG, 0);
+                // No need to increment the address because the logic does it automatically when MCA_WREITE_AACR_AUTOINC is set
+                fapi2::putScom(p, PT::RMW_WRT_BUF_DATA_REG, l_value_second);
+                // In order for the data to actually be written into the RMW buffer, we must issue a putscom to the MCA_AAER register
+                // This register is used for the ECC, we will just write all zero to this register.  The ECC will be auto generated
+                // when the aacr MCA_WREITE_AACR_ECCGEN bit is set
+                fapi2::putScom(p, PT::RMW_WRT_BUF_ECC_REG, 0);
+            }
+            l_aacr.template insertFromRight<PT::RMW_WRT_ADDRESS, PT::RMW_WRT_ADDRESS_LEN>(PT::MAINT_DATA_INDEX_END);
+        }
+    }
+}
+
+template< mss::mc_type MC = DEFAULT_MC_TYPE, fapi2::TargetType T, typename TT = mcbistTraits<MC, T> >
+inline fapi2::ReturnCode load_config( const fapi2::Target<T>& i_target, const mcbist::program<MC>& i_program )
+{
+    // Copy the program's config settings - we want to modify them if we're in sim.
+    fapi2::buffer<uint64_t> l_config = i_program.iv_config;
+    // If we're running in Cronus, there is no interrupt so any attention bits will
+    // hang something somewhere. Make sure there's nothing in this config which can
+    // turn on attention bits unless we're running in hostboot
+#ifndef __HOSTBOOT_MODULE
+    if(TT::CFG_ENABLE_ATTN_SUPPORT == mss::states::YES)
+    {
+        l_config.template clearBit<TT::CFG_ENABLE_HOST_ATTN>();
+        l_config.template clearBit<TT::CFG_ENABLE_SPEC_ATTN>();
+    }
+#endif
+    fapi2::putScom(i_target, TT::CFGQ_REG, l_config);
+}
+
+template<  mss::mc_type MC = DEFAULT_MC_TYPE, fapi2::TargetType T, typename TT = mcbistTraits<MC, T> >
+inline fapi2::ReturnCode load_fifo_mode( const fapi2::Target<T>& i_target, const mcbist::program<MC>& i_program )
+{
+    // Checks if FIFO mode is required by checking all subtests
+    const auto l_subtest_it = std::find_if(i_program.iv_subtests.begin(),
+                                           i_program.iv_subtests.end(),
+                                           []( const mss::mcbist::subtest_t<MC, T, TT>& i_rhs) -> bool
+                                           {
+                                               return i_rhs.fifo_mode_required();
```

Comment:
When is `fifo_mode_required`? Maybe we can skip all of this function if it happens to always be false.

Reply:
I'm not sure about that. The code checking it is hard to analyze, so I would leave it for now, and maybe revisit later